astropy/cosmology/units.py

```diff
@@ -255,16 +274,21 @@ def z_to_Tcmb(z):
     def Tcmb_to_z(T):
         return z_at_value(cosmology.Tcmb, T << u.K, **atzkw)
 
-    return u.Equivalency(
+    return Equivalency(
         [(redshift, u.K, z_to_Tcmb, Tcmb_to_z)],
         "redshift_temperature",
         {"cosmology": cosmology},
     )
 
 
 def with_redshift(
-    cosmology=None, *, distance="comoving", hubble=True, Tcmb=True, atzkw=None
-):
+    cosmology: Cosmology | str | None = None,
+    *,
+    distance: Literal["comoving", "lookback", "luminosity"] = "comoving",
+    hubble: bool = True,
+    Tcmb: bool = True,
+    atzkw: Mapping[str, Any] | None = None,
```

Comment:
I don't like that everything else here accepts `**atzkw`, but this one function requires `atzkw` to be a mapping instead. However, this pull request is correctly documenting the fact and we should try to keep type annotation pull requests separate from API changes, so for now we'll have to accept this as it is.